eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala

```diff
@@ -184,6 +184,38 @@ object Graph {
     shortestPaths.map(_.p).toSeq
   }
 
+  def breadthFirstSearch(g: DirectedGraph,
```

Comment:
Those are very good points. But there are also good reasons to use some of the heuristics we use for payments:

- channel are disabled by one side (the edge is missing in one direction) when one of the nodes is offline: when that happens we won't be able to route onion messages
- but we can't distinguish that case from the case where both nodes are online and the channel is just depleted on one side
- but this will only happen if there's a single channel between these two nodes and it is depleted, which indicates that this isn't a very busy/reliable edge
- so I guess we should consider those edges but give them a low score to use them only if we don't find better alternatives?
- large/old channels are a good indication that nodes are frequently online and responsive, so we'll want to favor them

We discussed that during yesterday's spec meeting, and the other implementations are either shoehorning into the existing path-finding, or still using dijkstra but with a different set of heuristics.

I don't think that introducing undirected edges and BFS is a very future-proof choice. Also, we've spent a lot of time fine-tuning the data structures used in our dijkstra implementation to make sure performance was good enough (in both CPU and memory terms). We'd need to spend a lot of time doing the same performance work on that BFS implementation if we want to go in that direction, which I'm not sure is worth the extra effort. What do you think of:

- adding node features to the graph: we've regularly realized that this would be necessary (for example for `var_onion_optin`, we just assumed support, but in practice it sometimes failed)
- create a copy of `dijkstraShortestPath` which is similar at the algorithm/data structure level but with a different scoring mechanism that we will tweak as we discover what actually makes sense in the real world

The graph data changes would look something like:

```
case class Vertex(features: Features[NodeFeature], edges: List[GraphEdge])
case class DirectedGraph(private val vertices: Map[PublicKey, Vertex])
```

This shouldn't require too much changes, since we're already reading all the node announcements when loading the `Router`. In `handleNodeAnnouncement` we'll need a mechanism to also update the `graph` with the latest node's features, but that shouldn't be too hard.